addons/l10n_it_edi/tests/test_ir_mail_server.py

```diff
@@ -0,0 +1,142 @@
+        cls.invoice = cls.env['account.move'].create({
+            'move_type': 'in_invoice',
+            'ref': '01234567890'
+        })
+        cls.attachment = cls.env['ir.attachment'].create({
+            'name': cls.invoice_filename1,
+            'datas': base64.encodebytes(cls.invoice_content),
+            'res_id': cls.invoice.id,
+            'res_model': 'account.move',
+        })
+        cls.edi_document = cls.env['account.edi.document'].create({
+            'edi_format_id': cls.edi_format.id,
+            'move_id': cls.invoice.id,
+            'attachment_id': cls.attachment.id,
+            'state': 'sent'
+        })
+
+        # Initialize the fetchmail server that has to be tested
+        cls.server = cls.env['fetchmail.server'].sudo().create({
+            'name': 'test_server',
+            'server_type': 'imap',
+            'l10n_it_is_pec': True})
+
+    @classmethod
+    def _get_test_content(cls, filename):
+        """ Get the content of a test file inside this module """
+        path = get_module_resource('l10n_it_edi', 'tests', 'expected_xmls', filename)
+        with tools.file_open(path, mode='rb') as test_file:
+            return test_file.read()
+
+    def _create_invoice(self, content, filename):
+        """ Create an invoice from given attachment content """
+        with patch.object(self.server._cr, 'commit', return_value=None):
+            return self.server._create_invoice_from_mail(content, filename, 'fake@address.be')
+
+    # -----------------------------
+    #
+    # Vendor bills
+    #
+    # -----------------------------
+
+    def test_receive_vendor_bill(self):
+        """ Test a sample e-invoice file from https://www.fatturapa.gov.it/export/documenti/fatturapa/v1.2/IT01234567890_FPR01.xml """
+        invoices = self._create_invoice(self.invoice_content, self.invoice_filename2)
+        self.assertTrue(bool(invoices))
+
+    def test_receive_same_vendor_bill_twice(self):
+        """ Test that the second time we are receiving a PEC mail with the same attachment, the second is discarded """
+        content = self.test_content.encode('utf-8')
+        for result in [True, False]:
+            invoice = self._create_invoice(content, self.invoice_filename2)
+            self.assertEqual(result, bool(invoice))
+
+    # -----------------------------
+    #
+    # Receipts
+    #
+    # -----------------------------
+
+    def _test_receipt(self, receipt_type, source_state, destination_state):
+        """ Test a receipt from the ones in the module's test files """
+
+        # Simulate the 'sent' state of the move, even if we didn't actually send an email in this test
+        self.invoice.l10n_it_send_state = source_state
+
+        # Create a fake receipt from the test file
+        receipt_filename = 'IT01234567890_FPR01_%s_001.xml' % receipt_type
+        receipt_content = self._get_test_content(receipt_filename).decode('utf-8')
```

Comment:
utf-8 is the default in Python 3, so any time you call `decode()` or `encode()` you don't need to specify it. Also, some of your decoding/encoding calls seem to be useless, like the b64 one. `ir.attachment` has a `raw` field that can be used to read/write the file contents without the b64 wrapping.

Reply:
Ok